### PR TITLE
tools to gracefully decommission node 

### PR DESCRIPTION
**Pull request by mzmssg**

### CODE REVIEW

**Comment by fanyangCS** [comment]
I remember you have a design doc. Can you include the design in the document?

**Comment by mzmssg** [comment]
@fanyangCS 

That's a service design to automate maintaining process. Put it here seems too excessive.

File: docs/tools/node_maintain.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+10.0.0.3 work-2
+10.0.0.3 work-3
+```
+
+At some point, `work-1` and `work-2` encounter hardware issues, you want to gracefully decommission them.
+
+##### 1. Edit the node list
+
+```bash
+python node_maintain.py {master_ip} node-list add 10.0.0.2,10.0.0.3
```
Comment (fanyangCS) [inline docs/tools/node_maintain.md]: what does this master_ip mean? Yarn RM? better explain it in more details.

File: tools/node_maintain.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+            break
+        else:
+            logger.info("Unready nodes: {}".format(unready_nodes))
+        logger.info("Waiting...")
+        time.sleep(30)
+    logger.info("Successfully refresh nodes.")
+
+def setup_parser():
+    top_parser = argparse.ArgumentParser()
+    top_parser.add_argument("ip", help="Master node ip")
```
Comment (fanyangCS) [inline tools/node_maintain.py]: change "Master node ip" to a more meaningful name

File: tools/node_maintain.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    nodelist_subparsers = nodelist_parser.add_subparsers(dest="action")
+
+    parser_get = nodelist_subparsers.add_parser("get", help="get unhealthy node list")
+    parser_get.set_defaults(func=get_decommission_nodes)
+
+    parser_add = nodelist_subparsers.add_parser("add", help="add unhealthy nodes")
+    parser_add.add_argument("nodes", help='support comma-delimited node list')
+    parser_add.set_defaults(func=add_decommission_nodes)
+
+    parser_remove = nodelist_subparsers.add_parser("remove", help="remove unhealthy nodes")
```
Comment (fanyangCS) [inline tools/node_maintain.py]: "remove" means decommission? better change it to "decommission", in help, you better say "graceful decommissioning nodes, jobs on the nodes will not be killed."

File: tools/node_maintain.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
+import requests
```
Comment (fanyangCS) [inline tools/node_maintain.py]: maybe /src/utilities is a better place to keep it?

File: docs/tools/node_maintain.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+```
+
+##### 2. Refresh nodes
+
+```bash
+python node_maintain.py {master_ip} refresh
```
Comment (fanyangCS) [inline docs/tools/node_maintain.md]: do you have any assumption on the execution path of node_maintain.py? If yes, please state so explicitly in the document.

File: tools/node_maintain.py (hunk 4 above)
Comment (ydye) [reply tools/node_maintain.py]: Should be in the path ```src/```

Comment (mzmssg) [reply tools/node_maintain.py]: `src/tools` seems OK. Then move it to `/deployment/paiLibrary` as integrated into paictl.

**Comment by fanyangCS** [comment]
ok. that is unnecessary. but better explain briefly what this tool does under the hood. this is a good example. [https://github.com/Microsoft/pai/blob/pai-0.9.y/docs/alerting/README.md](https://github.com/Microsoft/pai/blob/pai-0.9.y/docs/alerting/README.md)

---
In reply to: [463925013](https://github.com/Microsoft/pai/pull/2139#issuecomment-463925013) [](ancestors = 463925013)

File: docs/tools/node_maintain.md (hunk 1 above)
Comment (mzmssg) [reply docs/tools/node_maintain.md]: add more details

File: tools/node_maintain.py (hunk 2 above)
Comment (mzmssg) [reply tools/node_maintain.py]: add more details in parser

File: tools/node_maintain.py (hunk 3 above)
Comment (mzmssg) [reply tools/node_maintain.py]: Nope, refined parser helper

File: docs/tools/node_maintain.md (hunk 5 above)
Comment (mzmssg) [reply docs/tools/node_maintain.md]: added

**Comment by mzmssg** [comment]
@fanyangCS I have update more details in doc

File: tools/node_maintain.py (hunk 4 above)
Comment (mzmssg) [reply tools/node_maintain.py]: moved to `src/tools`

File: docs/tools/node_maintain.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
+# node_maintain
+
+## Overview
+
+Graceful decommissioning could minimize the impact on existing job when remove unhealthy nodes.
```
Comment (fanyangCS) [inline docs/tools/node_maintain.md]: >remove [](start = 72, length = 6)

removing

File: docs/tools/node_maintain.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+## Commands
+
+Currently, node-list update and component refresh are both manual actions. 
+This tool provides corresponding command for them. 
+The only mandatory argument for all commands is `master_ip`. Usually, components share the same master ip,
+in different case(i.e. load-balance), you could overwrite `master_ip` separately by `--kubernetes-ip` and `--yarn-ip`
+
+So, a common invoking looks like:
+```bash
+python node_maintain.py {master_ip} sub-command sub-command-arguments
```
Comment (fanyangCS) [inline docs/tools/node_maintain.md]: >node_maintain.p [](start = 7, length = 15)

does the command has to be run in the path /src/tools? If so, please state it in the manual.

File: docs/tools/node_maintain.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+Components could sync the list and decommission them as needed.
+
+**NOTES: For now, only YARN nodes are decommissioned, HDFS doesn't.**
+
+## Commands
+
+Currently, node-list update and component refresh are both manual actions. 
+This tool provides corresponding command for them. 
+The only mandatory argument for all commands is `master_ip`. Usually, components share the same master ip,
+in different case(i.e. load-balance), you could overwrite `master_ip` separately by `--kubernetes-ip` and `--yarn-ip`
```
Comment (fanyangCS) [inline docs/tools/node_maintain.md]: >kubernetes [](start = 87, length = 10)

kubernetes-ip is the api-server's ip? and yarn-ip is the RM's ip? 
Please explain it more clearly.

File: docs/tools/node_maintain.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+For convenience, in this section, we assume you have a cluster without load-balance, as below:
+```
+10.0.0.10 master
+10.0.0.1 work-1
+10.0.0.2 work-2
+10.0.0.3 work-3
+```
+
+### Decommission nodes
```
Comment (fanyangCS) [inline docs/tools/node_maintain.md]: >Decommission [](start = 4, length = 12)

Please add some doc on how to recommission a node

File: docs/tools/node_maintain.md (hunk 8 above)
Comment (mzmssg) [reply docs/tools/node_maintain.md]: Renamed and add ip source

File: docs/tools/node_maintain.md (hunk 9 above)
Comment (mzmssg) [reply docs/tools/node_maintain.md]: I think this example is enough?
https://github.com/Microsoft/pai/blob/859f291c0fc1babe54e643d54c4859f057ffafc5/docs/tools/node_maintain.md#recommission-nodes

**Pull request merged**